### DISCUSSION TITLE
[Feature Request] Add first-party router support in middleware

### DISCUSSION THREAD

**Discussion by SukkaW** [vercel/next.js#37474] [Ideas]
## Context

- Since `next@12.1.7-canary.9` Next.js uses one single root middleware instead of nested middleware, which means developers have to declare all middleware logic in one place.
- After https://github.com/vercel/next.js/pull/37121, Next.js will also execute middleware on internal requests, which might degrade the performance if some heavy logic is being added in the middleware (Adding a `100ms` delay to every `/_next/data` request sounds like a bad idea, especially if developers forget to set a proper `config.matcher`).
- https://github.com/vercel/next.js/pull/37177 allows the middleware to only execute on specific routes, but developers still have to implement their own router in case multiple logic have to be added:

```js
import { NextResponse } from 'next/server';

export const config = {
  matcher: [
    '/user/:path*', // Access user control panel requires extra authentication (E.g. SSO based),
    '/register*', // Add extra geo country restriction on register page,
    '/blocked' // We do not allow blocked page to be accessed directly 
  ]
}

export default (req) => {
  if (req.nextUrl.pathname.startsWith('/user')) {
    // Add extra SSO OAuth based authentication here
  } else if (req.nextUrl.pathname.startsWith('/register')) {
    if (req.geo.country === 'T1') {
      return NextResponse.rewrites('/blocked');
    }
  } else if (req.nextUrl.pathname.startsWith('/blocked')) {
    return NextResponse.rewrites('/404');
  }
  return NextResponse.next();
}
```

### Describe the solution you'd like

Imagine if we have a named export `NextMiddlewareRouter` from `next/server`:

```js
import { NextResponse, NextMiddlewareRouter } from 'next/server';

export const config = {
  matcher: [
    '/user/:path*', // Access user control panel requires extra authentication (E.g. SSO based),
    '/register*', // Add extra geo country restriction on register page,
    '/blocked' // We do not allow blocked page to be accessed directly 
  ]
}

const router = new NextMiddlewareRouter({
  // maybe some options?
});

// Maybe in the future we can statically analyze the regexp and calculate the matcher?
router.get('/user/:path*', (req) => {
  // Add extra SSO OAuth based authentication here
});
router.get('/register*', (req) => {
  if (req.geo.country === 'T1') {
    return NextResponse.rewrites('/blocked');
  }
});
router.get('/blocked*', (req) => {
  return NextResponse.rewrites('/404');
});

export default (req) => router(req);
```

### Describe alternatives you've considered

Let the community implements a home-baked router implementation.

**Comment by Kikobeats** [comment]
Most of express-like routers (being [router](https://github.com/pillarjs/router) one of the most popular) are based on regex pattern matching (using [path-to-regexp](https://github.com/pillarjs/path-to-regexp) or specific data structures).

Also, you can find some libraries published at npm aiming to be edge-ready, but they are still based on the same concept:

- https://github.com/kwhitley/itty-router
- https://github.com/honojs/hono

There is nothing bad with that, but ideally we can take advantage of the [URLPattern](https://web.dev/urlpattern/) since it's a built-in API available in the middleware context:

```js
// middleware.ts
import { NextRequest, NextResponse } from https://github.com/vercel/next.js

const PATTERNS = [
  [
    new URLPattern({ pathname: '/:locale/:slug' }),
    ({ pathname }) => pathname.groups,
  ],
]

const params = (url) => {
  const input = url.split('?')[0]
  let result = {}

  for (const [pattern, handler] of PATTERNS) {
    const patternResult = pattern.exec(input)
    if (patternResult !== null && 'pathname' in patternResult) {
      result = handler(patternResult)
      break
    }
  }
  return result
}

// rewrite `{domain}/{locale}/{slug}` into `{locale}.{domain}/{slug}`
export function middleware(request: NextRequest) {
  const { locale, slug } = params(request.url)

  if (locale && slug) {
    const { search, protocol, host } = request.nextUrl
    const url = new URL(`${protocol}//${locale}.${host}/${slug}${search}`)
    return NextResponse.redirect(url)
  }
}
```

I feel it could be wrapper by a top level API, so you wouldn't need to deal with URLPattern details.
